### PR TITLE
Riscv

### PR DESCRIPTION
**Pull request by nastya-nizharadze**
Add context switch

### CODE REVIEW

File: src/arch/riscv/kernel/context.h [hunk 1]
```diff
@@ -1,7 +1,35 @@
...
+#ifndef __ASSEMBLER__
+
 struct context {
+	unsigned int ra;
+	unsigned int sp;
+	unsigned int s[12];
+	unsigned int mstatus;
 };
 
+#else
```
Comment (anton-bondarev) [inline src/arch/riscv/kernel/context.h]: #endif ?
Do you want to hide definitions in none assembly code?

File: templates/riscv/qemu-sifive-u/mods.config [hunk 2]
```diff
@@ -21,5 +22,9 @@ configuration conf {
 	include embox.compat.libc.math_builtins
 	include embox.kernel.spinlock(spin_debug = false)
 
+	include embox.kernel.task.resource.errno
+
 	include embox.test.stdlib.setjmp_test
+	include embox.test.kernel.timer_test
+	include embox.test.hal.context_switch_test(stack_sz=0x40)
```
Comment (anton-bondarev) [inline templates/riscv/qemu-sifive-u/mods.config]: Maybe you should switch on more multi-threading features?

File: src/arch/riscv/kernel/context.c [hunk 3]
```diff
@@ -1,11 +1,15 @@
-#include <kernel/panic.h>
+#include <stdint.h>
 #include <hal/context.h>
+#include <hal/ipl.h>
+#include <asm/regs.h>
 
 void context_init(struct context *ctx, unsigned int flags,
 		void (*routine_fn)(void), void *sp) {
+	ctx->sp = (uint32_t) sp;
```
Comment (Deryugin) [inline src/arch/riscv/kernel/context.c]: `ctx->sp` is `unsigned int`, but you cast `sp` to `(uint32_t)`, I recommend to use `uint32_t` explicitly for `ctx` fields (if they are 32 bit wide).

File: src/arch/riscv/kernel/context.c [hunk 4]
```diff
@@ -1,11 +1,15 @@
-#include <kernel/panic.h>
+#include <stdint.h>
 #include <hal/context.h>
+#include <hal/ipl.h>
+#include <asm/regs.h>
 
 void context_init(struct context *ctx, unsigned int flags,
 		void (*routine_fn)(void), void *sp) {
```
Comment (Deryugin) [inline src/arch/riscv/kernel/context.c]: I think that unused fields of `ctx` should be set to zero in this function to make sure there are no "garbage" in the register.
You can use `memset(ctx, 0, sizeof(*ctx))` at the beginning of the function.

File: src/arch/riscv/kernel/context_switch.S [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	sw      s7, CTX_S7(a0)
+	sw      s8, CTX_S8(a0)
+	sw      s9, CTX_S9(a0)
+	sw      s10, CTX_S10(a0)
+	sw      s11, CTX_S11(a0)
+	csrr    t6, mstatus
+	sw      t6, CTX_MSTATUS(a0)
+	
+	lw      sp, CTX_SP(a1)
+	nop
```
Comment (Deryugin) [inline src/arch/riscv/kernel/context_switch.S]: Are these `nop`s really required for correct context switch?

File: src/arch/riscv/kernel/context.h (hunk 1 above)
Comment (nastya-nizharadze) [reply src/arch/riscv/kernel/context.h]: Not really. Fixup done.

File: templates/riscv/qemu-sifive-u/mods.config (hunk 2 above)
Comment (nastya-nizharadze) [reply templates/riscv/qemu-sifive-u/mods.config]: I think I'll do it later, after implementing vfork.

File: src/arch/riscv/kernel/context_switch.S (hunk 5 above)
Comment (nastya-nizharadze) [reply src/arch/riscv/kernel/context_switch.S]: As it turned out, in RISC-V there is no slot delay, nor can the data hazard occur, so no.

**Comment by anton-bondarev** [comment]
Work on #1047

**Pull request merged**